sdk/core/azure-core/test/ut/transport_adapter.cpp

```diff
@@ -26,6 +26,88 @@ namespace Azure { namespace Core { namespace Test {
   Azure::Core::Http::HttpPipeline TransportAdapter::pipeline(policies);
   Azure::Core::Context TransportAdapter::context = Azure::Core::GetApplicationContext();
 
+  // multiThread test requires `ConnectionsOnPool` hook which is only available when building
+  // TESTING_BUILD. This test cases are only built when that case is true.`
+  TEST_F(TransportAdapter, getMultiThread)
+  {
+    std::string host("http://httpbin.org/get");
+    Azure::Core::Http::CurlConnectionPool::ClearIndex();
+
+    auto threadRoutine = [host]() {
+      auto request = Azure::Core::Http::Request(Azure::Core::Http::HttpMethod::Get, host);
+      auto response = pipeline.Send(context, request);
+      checkResponseCode(response->GetStatusCode());
+      auto expectedResponseBodySize = std::stoull(response->GetHeaders().at("content-length"));
+      CheckBodyFromBuffer(*response, expectedResponseBodySize);
+    };
+
+    std::thread t1(threadRoutine);
+    std::thread t2(threadRoutine);
+    t1.join();
+    t2.join();
+
+    // 2 connections must be available at this point
+    EXPECT_EQ(Http::CurlConnectionPool::ConnectionsOnPool("httpbin.org"), 2);
+
+    std::thread t3(threadRoutine);
+    std::thread t4(threadRoutine);
+    std::thread t5(threadRoutine);
+    t3.join();
+    t4.join();
+    t5.join();
+    // wait a few ms for connections to go back to pool.
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
```

Comment:
To me, it looks like the only real change being made here is this line got moved from before `EXPECT_EQ(Http::CurlConnectionPool::ConnectionsOnPool("httpbin.org"), 2);` to after it.

`std::this_thread::sleep_for(std::chrono::milliseconds(500));`

Is that correct? Can you please explain why.

Why did we increase it from 500 ms to 1000? Is that necessary?

Reply:
Generally speaking, relying on timing for tests to pass is going to be brittle.

Reply:
500 ms could be not enough to to make sure connections were moved back to pool.
Whenever each of the previous threads finishes, it calls the destructor from the curl session which will move the connection back to the pool.  For moving the connection, each thread needs to lock a mutex, so they need to wait and do it one after the other.
By waiting 500 ms, sometimes CI was failing because by the time we check how many connection are in the pool, only 1 or 2 were moved and the other one was still pending to be moved.